pkg/p2p/kadcast/packet.go

```diff
@@ -0,0 +1,252 @@
+		payload: payload,
+	}
+}
+
+// -------- General Packet De/Serialization tools -------- //
+
+// Gets a stream of bytes and slices it between headers of Kadcast
+// protocol and the payload.
+func getPacketFromStream(stream []byte) Packet {
+	var headers [24]byte
+	copy(headers[:], stream[0:23])
+	return Packet{
+		headers: headers,
+		payload: stream[23:],
+	}
+}
+
+// Deserializes the packet into an slice of bytes.
+func (pac Packet) asBytes() []byte {
+	return append(pac.headers[:], pac.payload[:]...)
+}
+
+// Returns the headers info sliced into three pieces:
+// Packet type, SenderId, IdNonce and senderPort.
+func (pac Packet) getHeadersInfo() (byte, [16]byte, [4]byte, [2]byte) {
+	// Construct type, senderID and Nonce
+	typ := pac.headers[0]
+	var senderID [16]byte
+	copy(senderID[:], pac.headers[1:17])
+	var nonce [4]byte
+	copy(nonce[:], pac.headers[17:21])
+	var peerPort [2]byte
+	copy(peerPort[:], pac.headers[21:23])
+	return typ, senderID, nonce, peerPort
+}
+
+// Gets the Packet headers parts and puts them into the
+// header attribute of the Packet.
+func (pac *Packet) setHeadersInfo(tipus byte, router Router, destPeer Peer) {
+	var headers []byte
+	// Add `Packet` type.
+	headers = append(headers[:], tipus)
+	// Add MyPeer ID
+	headers = append(headers[:], router.MyPeerInfo.id[:]...)
+	// Attach IdNonce
+	idNonce := getBytesFromUint32(router.myPeerNonce)
+	headers = append(headers[:], idNonce[:]...)
+	// Attach Port
+	port := getBytesFromUint16(destPeer.port)
+	headers = append(headers[:], port[:]...)
+
+	// Build headers array from the slice.
+	var headersArr [24]byte
+	copy(headersArr[:], headers[0:24])
+
+	pac.headers = headersArr
+}
+
+// -------- NODES Packet De/Serialization tools -------- //
+
+// Builds the payload of a `NODES` message by collecting,
+// deserializing and adding to the packet's payload the
+// peerInfo of the `K` closest Peers in respect to a certain
+// target Peer.
+func (pac *Packet) setNodesPayload(router Router, targetPeer Peer) int {
+	// Get `K` closest peers to `targetPeer`.
+	kClosestPeers := router.getXClosestPeersTo(K, targetPeer)
+	// Compute the ammount of Peers that will be sent and add it
+	// as a two-byte array.
+	count := getBytesFromUint16(uint16(len(kClosestPeers)))
+	pac.payload = append(pac.payload[:], count[:]...)
+	// Serialize the Peers to get them in `wire-format`,
+	// basically, represented as bytes.
+	for _, peer := range kClosestPeers {
+		pac.payload = append(pac.payload[:], peer.deserializePeer()...)
+	}
+	return len(kClosestPeers)
+}
+
+// Analyzes if the announced number of Peers included on the
+// `NODES` message payload is the same as the recieved one.
+// Returns `true` if it is correct and `false` otherways.
+func (pac Packet) checkNodesPayloadConsistency(byteNum int) bool {
+	// Get number of Peers announced.
+	peerNum := binary.BigEndian.Uint16(pac.payload[0:2])
+	// Get peerSlice length subtracting headers and count.
+	peerSliceLen := byteNum - 26
```

Comment:
Even the comment describes it, I did:
```
peerSliceLen := byteNum - (len(pac.headers) + 2)
```
since the peerCount len cannot be obtained from anwhere (it does not make sense to do lots of stuff at least to get it as a variable).